### PR TITLE
Refactor fileRequester & complete fileSender implementation

### PR DESCRIPTION
**Pull request by jaceko**

### CODE REVIEW

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java [hunk 1]
```diff
@@ -60,24 +57,24 @@ public Response post(final String uri, final MediaType mediaType, final Multival
      * @return the response from the Http request.
      */
     public Response get(final String uri, final MediaType mediaType, final MultivaluedHashMap<String, Object> headers) {
-        return getClient()
+        return alfrescoRequestWith(uri, mediaType, headers).get();
+    }
+
+    private Invocation.Builder alfrescoRequestWith(final String uri, final MediaType mediaType, final MultivaluedHashMap<String, Object> headers) {
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java]: Can we static import the builder?

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java [hunk 2]
```diff
@@ -60,24 +57,24 @@ public Response post(final String uri, final MediaType mediaType, final Multival
...
-                .headers(headers)
-                .get();
+                .headers(headers);
     }
 
-    private Client getClient() {
-        Client client;
-
+    private Client client() {
         if ("none".equals(proxyType)) {
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java]: What do you think about having none as a static NO_PROXY ?

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/Headers.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import static java.util.Collections.singletonList;
+
+import javax.ws.rs.core.MultivaluedHashMap;
+
+public class Headers {
+
+    public static final String ALFRESCO_USER_ID = "cppuid";
+
+    public static MultivaluedHashMap<String, Object> headersWithUserId(String id) {
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/Headers.java]: final parameter! yay! :D

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java [hunk 4]
```diff
@@ -29,21 +32,23 @@
...
-
-        //TODO: build response
-
-        return null;
+        try {
+            final Response response = restClient.post(alfrescoUploadPath, MULTIPART_FORM_DATA_TYPE, headersWithUserId(alfrescoUploadUser), entity(content, MULTIPART_FORM_DATA_TYPE));
+            final String responseEntity = response.readEntity(String.class);
+            if (response.getStatusInfo() != OK || isEmpty(responseEntity)) {
+                //Alfresco is *very* accepting - failed exceptions represent service outages/problems only.
+                throw new FileOperationException(String.format("Error while uploading document. Code:%d, Reason:%s", response.getStatus(), response.getStatusInfo().getReasonPhrase()));
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java]: static import format

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java [hunk 5]
```diff
@@ -29,21 +32,23 @@
...
-
-        restClient.post(alfrescoUploadPath, MULTIPART_FORM_DATA_TYPE, headers, entity(content, MULTIPART_FORM_DATA_TYPE));
-
-        //TODO: build response
-
-        return null;
+        try {
+            final Response response = restClient.post(alfrescoUploadPath, MULTIPART_FORM_DATA_TYPE, headersWithUserId(alfrescoUploadUser), entity(content, MULTIPART_FORM_DATA_TYPE));
+            final String responseEntity = response.readEntity(String.class);
+            if (response.getStatusInfo() != OK || isEmpty(responseEntity)) {
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java]: maybe add a couple of blank lines here for formatting, to separate request, check etc

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java [hunk 6]
```diff
@@ -29,21 +32,23 @@
...
+                //Alfresco is *very* accepting - failed exceptions represent service outages/problems only.
+                throw new FileOperationException(String.format("Error while uploading document. Code:%d, Reason:%s", response.getStatus(), response.getStatusInfo().getReasonPhrase()));
+            }
+            final Object responseDocument = defaultConfiguration().jsonProvider().parse(responseEntity);
+            final String alfrescoAssetId = JsonPath.read(responseDocument, "$.nodeRef").toString().replace("workspace://SpacesStore/", "");
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java]: static import?

File: file/file-alfresco/src/test/java/uk/gov/justice/services/file/alfresco/requester/AlfrescoFileRequesterIT.java [hunk 7]
```diff
@@ -26,93 +25,86 @@
...
-               .withHeader(ID, equalTo("user1234")));
+        verify(getRequestedFor(urlEqualTo("/alfresco" + ALFRESCO_WORKSPACE_PATH + fileId + "/content/" + fileName))
+                .withHeader("cppuid", equalTo("user1234")));
     }
 
     @Test
-    public void shouldReturnResponseFromAlfresco(){
+    public void shouldReturnResponseFromAlfresco() {
         final String fileId = randomUUID().toString();
         final String mimeType = "text/plain";
```
Comment (matt-rich) [inline file/file-alfresco/src/test/java/uk/gov/justice/services/file/alfresco/requester/AlfrescoFileRequesterIT.java]: perhaps make these common test data values constants as they are repeated ?

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java [hunk 8]
```diff
@@ -29,21 +32,23 @@
...
+            final String mimeType = JsonPath.read(responseDocument, "$.fileMimeType");
```
Comment (jaceko) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java]: using constants for json paths would obfuscate the code a little, I usually prefer to have them directly in code, if they are not reused elsewhere.

extracting private method done

File: file/file-alfresco/src/test/java/uk/gov/justice/services/file/alfresco/requester/AlfrescoFileRequesterIT.java (hunk 7 above)
Comment (jaceko) [inline file/file-alfresco/src/test/java/uk/gov/justice/services/file/alfresco/requester/AlfrescoFileRequesterIT.java]: I'm not sure if it's a good practice to use constants in tests.
I think it makes tests more readable if you directly declare and assign variables in each individual test case which you later use in assertions

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java (hunk 6 above)
Comment (jaceko) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/sender/AlfrescoFileSender.java]: done

File: file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java [hunk 9]
```diff
@@ -19,12 +21,13 @@
 @ApplicationScoped
 public class AlfrescoRestClient {
 
+    public static final String NO_PROXY = "none";
```
Comment (matt-rich) [inline file/file-alfresco/src/main/java/uk/gov/justice/services/file/alfresco/common/AlfrescoRestClient.java]: private ?

**Pull request merged**